File: mcp-server/localSkills.mjs
/**
 * Local-skill check — Node port of integrations/local_skills.py.
 *
 * Don't recommend skills the user already has installed. We scan the known
 * Claude Code skill locations, read each SKILL.md's frontmatter `name`, and use
 * those names to filter / demote federation candidates.
 *
 * Scanned locations (Claude Code; v1 scope):
 *   ~/.claude/skills/<skill>/SKILL.md          (global)
 *   <cwd>/.claude/skills/<skill>/SKILL.md       (project)
 *   $CLAUDE_PROJECT_DIR/.claude/skills/...       (Claude Code sets this for MCP servers)
 *
 * Matching is by normalized skill `name` (the spec's unique id within a scope).
 */

import fs from "node:fs";
import os from "node:os";
import path from "node:path";

// Mirror local_skills.py: _FM_NAME = ^---\n ... \bname\s*:\s*([^\n]+) ... \n---
const FM_NAME = /^---\s*\n[\s\S]*?\bname\s*:\s*([^\n]+)[\s\S]*?\n---/;
const TOK = /[a-z0-9]+/g;

function normName(s) {
  return (String(s).toLowerCase().match(TOK) || []).join("-");
}

function readName(skillMd) {
  let text;
  try {
    const fd = fs.openSync(skillMd, "r");
    try {
      const buf = Buffer.alloc(4000);
      const n = fs.readSync(fd, buf, 0, 4000, 0);
      text = buf.toString("utf8", 0, n);
    } finally {
      fs.closeSync(fd);
    }
  } catch {
    return null;
  }
  const m = FM_NAME.exec(text);
  if (m) return m[1].trim().replace(/^['"]|['"]$/g, "");
  // fallback: parent dir name
  return path.basename(path.dirname(skillMd));
}

function candidateDirs() {
  const home = os.homedir();
  const cwd = process.cwd();
  const proj = process.env.CLAUDE_PROJECT_DIR || cwd;
  const roots = [
    path.join(home, ".claude", "skills"),
    path.join(cwd, ".claude", "skills"),
    path.join(proj, ".claude", "skills"),
  ];
  const seen = new Set();
  const out = [];
  for (const r of roots) {
    if (!seen.has(r)) {
      seen.add(r);
      out.push(r);
    }
  }
  return out;
}

function listDirs(root) {
  try {
    return fs
      .readdirSync(root, { withFileTypes: true })
      .filter((d) => d.isDirectory())
      .map((d) => d.name);
  } catch {
    return [];
  }
}

function listMdFiles(root) {
  try {
    return fs
      .readdirSync(root, { withFileTypes: true })
      .filter((d) => d.isFile() && d.name.toLowerCase().endsWith(".md"))
      .map((d) => d.name);
  } catch {
    return [];
  }
}

/** Set of normalized names of locally-installed skills. */
export function installedSkillNames(extraDirs = []) {
  const names = new Set();
  const dirs = [...candidateDirs(), ...extraDirs];
  for (const root of dirs) {
    if (!fs.existsSync(root) || !fs.statSync(root).isDirectory()) continue;
    // <root>/<skill>/SKILL.md layout
    for (const sub of listDirs(root)) {
      const md = path.join(root, sub, "SKILL.md");
      if (fs.existsSync(md)) {
        const nm = readName(md);
        if (nm) names.add(normName(nm));
      }
    }
    // also bare <root>/<name>.md layout
    for (const file of listMdFiles(root)) {
      if (file.toLowerCase() === "readme.md") continue;
      const nm = readName(path.join(root, file));
      if (nm) names.add(normName(nm));
    }
  }
  return names;
}

/**
 * Split candidates into [newOnes, alreadyInstalled] by normalized name.
 * newOnes are the ones to recommend; alreadyInstalled the ones to optionally
 * note ("you already have this").
 */
export function filterCandidates(candidates, installed) {
  if (installed === undefined) installed = installedSkillNames();
  const newOnes = [];
  const have = [];
  for (const c of candidates) {
    if (installed.has(normName(c.name || ""))) have.push(c);
    else newOnes.push(c);
  }
  return [newOnes, have];
}

export { normName };

File: integrations/local_skills.py
"""Local-skill check — don't recommend skills the user already has installed.

Release-critical (Bennett flagged): the agent shouldn't pull a skill that's
already present. We scan the known Claude Code skill locations, read each
SKILL.md's `name`, and use those names to filter / demote federation candidates.

Scanned locations (Claude Code; v1 scope):
  ~/.claude/skills/<skill>/SKILL.md          (global)
  <cwd>/.claude/skills/<skill>/SKILL.md       (project)
  $CLAUDE_PROJECT_DIR/.claude/skills/...       (if set by the hook env)

Matching is by normalized skill `name` (the spec's unique id within a scope),
with a description-similarity fallback so near-identical re-publishes are caught
even when names differ slightly.
"""
from __future__ import annotations

import os
import re
import glob

_FM_NAME = re.compile(r"^---\s*\n.*?\bname\s*:\s*([^\n]+).*?\n---", re.DOTALL)
_TOK = re.compile(r"[a-z0-9]+")


def _norm_name(s: str) -> str:
    return "-".join(_TOK.findall(s.lower()))


def _read_name(skill_md: str) -> str | None:
    try:
        with open(skill_md, "r", encoding="utf-8", errors="replace") as f:
            text = f.read(4000)
    except OSError:
        return None
    m = _FM_NAME.match(text)
    if m:
        return m.group(1).strip().strip("'\"")
    # fallback: parent dir name
    return os.path.basename(os.path.dirname(skill_md))


def _candidate_dirs() -> list[str]:
    home = os.path.expanduser("~")
    cwd = os.getcwd()
    proj = os.environ.get("CLAUDE_PROJECT_DIR", cwd)
    roots = [
        os.path.join(home, ".claude", "skills"),
        os.path.join(cwd, ".claude", "skills"),
        os.path.join(proj, ".claude", "skills"),
    ]
    # de-dup while preserving order
    seen, out = set(), []
    for r in roots:
        if r not in seen:
            seen.add(r)
            out.append(r)
    return out


def installed_skill_names(extra_dirs: list[str] | None = None) -> set[str]:
    """Return the set of normalized names of locally-installed skills."""
    names: set[str] = set()
    dirs = _candidate_dirs() + list(extra_dirs or [])
    for root in dirs:
        if not os.path.isdir(root):
            continue
        for md in glob.glob(os.path.join(root, "*", "SKILL.md")):
            nm = _read_name(md)
            if nm:
                names.add(_norm_name(nm))
        # also bare <root>/<name>.md layout
        for md in glob.glob(os.path.join(root, "*.md")):
            if os.path.basename(md).lower() == "readme.md":
                continue
            nm = _read_name(md)
            if nm:
                names.add(_norm_name(nm))
    return names


def filter_candidates(candidates: list[dict], installed: set[str] | None = None
                      ) -> tuple[list[dict], list[dict]]:
    """Split candidates into (new, already_installed) by normalized name.

    `candidate['name']` is the federation skill name. Returns the new ones first
    (to recommend) and the already-installed ones (to optionally note 'you
    already have this').
    """
    if installed is None:
        installed = installed_skill_names()
    new, have = [], []
    for c in candidates:
        if _norm_name(c.get("name", "")) in installed:
            have.append(c)
        else:
            new.append(c)
    return new, have
